src/bvals/cc/sparse_flux_correction_cc_in_one.cpp

```diff
@@ -0,0 +1,335 @@
+          const int l = loop_idx / NmNnNkNjNi;
+          const int m = (loop_idx % NmNnNkNjNi) / NnNkNjNi;
+          const int n = (loop_idx % NnNkNjNi) / NkNjNi;
+          const int ck = (loop_idx % NkNjNi) / NjNi;
+          const int cj = (loop_idx % NjNi) / ni;
+          const int ci = loop_idx % ni;
+
+          const int k = ks + 2 * ck;
+          const int j = js + 2 * cj;
+          const int i = is + 2 * ci;
+
+          // For the given set of offsets, etc. this should work for any
+          // dimensionality since the same flux will be included multiple times
+          // in the average
+          const Real area00 = coords.Area(dir, k, j, i);
+          const Real area01 = coords.Area(dir, k, j + joff, i + ioff);
+          const Real area10 = coords.Area(dir, k + koff, j + joff, i);
+          const Real area11 = coords.Area(dir, k + koff, j, i + ioff);
+
+          Real avg_flx = area00 * flx(l, m, n, k, j, i);
+          avg_flx += area01 * flx(l, m, n, k + koff, j + joff, i);
+          avg_flx += area10 * flx(l, m, n, k, j + joff, i + ioff);
+          avg_flx += area11 * flx(l, m, n, k + koff, j, i + ioff);
+
+          avg_flx /= area00 + area01 + area10 + area11;
+          const int idx = ci + ni * (cj + nj * (ck + nk * (n + nn * (m + nm * l))));
+          buf_arr(idx) = avg_flx;
+        });
+
+    // Send the buffer
+    PARTHENON_REQUIRE(buf.GetState() == BufferState::stale, "Not sure how I got here.");
+#ifdef MPI_PARALLEL
+    Kokkos::fence();
+#endif
+    buf.Send();
+    return LoopControl::cont;
+  });
+
+  Kokkos::Profiling::popRegion();
+  return TaskStatus::complete;
+}
+
+TaskStatus StartReceiveSparseFluxCorrectionBuffers(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_ReceiveFluxCorrectionBuffers");
+  Mesh *pmesh = md->GetMeshPointer();
+  ForEachBoundary<BoundaryType::reflux_recv>(
+      md, [&](sp_mb_t pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
+        PARTHENON_DEBUG_REQUIRE(
+            pmesh->boundary_comm_reflux_map.count(ReceiveKey(pmb, nb, v)) > 0,
+            "Boundary communicator does not exist");
+        auto &buf = pmesh->boundary_comm_reflux_map[ReceiveKey(pmb, nb, v)];
+        buf.TryStartReceive();
+      });
+  Kokkos::Profiling::popRegion();
+  return TaskStatus::complete;
+}
+
+TaskStatus ReceiveSparseFluxCorrectionBuffers(std::shared_ptr<MeshData<Real>> &md) {
+  Kokkos::Profiling::pushRegion("Task_ReceiveFluxCorrectionBuffers");
+  bool all_received = true;
+  Mesh *pmesh = md->GetMeshPointer();
+  ForEachBoundary<BoundaryType::reflux_recv>(
+      md, [&](sp_mb_t pmb, sp_mbd_t rc, nb_t &nb, const sp_cv_t v) {
+        PARTHENON_DEBUG_REQUIRE(
+            pmesh->boundary_comm_reflux_map.count(ReceiveKey(pmb, nb, v)) > 0,
+            "Boundary communicator does not exist");
+        auto &buf = pmesh->boundary_comm_reflux_map[ReceiveKey(pmb, nb, v)];
+        all_received = all_received && buf.TryReceive();
```

Comment:
Not suggesting a change here, but I was wondering if there might be a performance gain by removing the Keys from the map (or a from a copy thereof) so that a) only buffers not received will be checked in the loop, and b) the lookup itself should become increasingly faster.
Though, probably not a performance bottleneck anyway.

Reply:
`TryReceive` should essentially just be a comparison in same rank communication or after an MPI message has been successfully received, so I don't think there is much performance to be gained by removing some of them. We would definitely get a little bit better performance if we used the boundary iteration order to turn the map into a vector like is done in the bvals in one code, but I think that is something to be added when/if the flux corrections are made to be *in one*.